### PR TITLE
rust: Update to 1.58.1

### LINKED ISSUE #7870: [Bug]: Cargo Clippy broken in Rust 1.56
### Problem description

Cargo clippy does not work after updating the rust package from v1.53 to v1.56. 
Clippy always fails because of an "internal compiler error". 
Other commands, such as cargo build, cargo run, cargo new and cargo fmt work as intended.
Command was working as intended before update. The same code now fails to be linted and compile. 
Uninstalling and reinstalling the rust package did not fix the problem.
Switching editions to 2021 inside Cargo.toml did not fix the problem.
Creating a brand new project and copy-pasting just the code from their github repo (in case project cache was somehow messed up) did not fix the problem.

Project 1:
```
Checking matrix v0.1.0 (/data/data/com.termux/files/home/projects/rust/matrix)
error: internal compiler error: compiler/rustc_mir_build/src/build/mod.rs:842:17: no terminator on block 4
  --> src/matrix.rs:24:5
   |
24 | /     pub fn identity(rows: usize, columns: usize) -> Self {
25 | |         let mut identity = Self::zeroes(rows, columns);
26 | |         let min = if rows < columns { rows } else { columns };
27 | |         for index in 0..min {
...  |
30 | |         identity
31 | |     }
   | |_____^

thread 'rustc' panicked at 'Box<dyn Any>', /home/builder/.termux-build/rust/src/compiler/rustc_errors/src/lib.rs:1093:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.1.56

query stack during panic:
#0 [mir_built] building MIR for `matrix::<impl at src/matrix.rs:10:1: 63:2>::identity`
#1 [unsafety_check_result] unsafety-checking `matrix::<impl at src/matrix.rs:10:1: 63:2>::identity`
end of query stack
error: could not compile `matrix`
```

Project 2:
```
Checking polynomial v0.1.0 (/data/data/com.termux/files/home/projects/rust/polynomial)
error: internal compiler error: compiler/rustc_mir_build/src/build/scope.rs:1397:17: cannot unwind from goto -> bb68
  --> src/polynomial/plot.rs:37:21                                   |
37 |                     print!(" ")                                 |                     ^^^^^^^^^^^
   |
   = note: this error: internal compiler error originates in the macro `$crate::format_args` (in Nightly builds, run with -Z macro-backtrace for more info)

thread 'rustc' panicked at 'Box<dyn Any>', /home/builder/.termux-build/rust/src/compiler/rustc_errors/src/lib.rs:1093:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.1.56

query stack during panic:
#0 [mir_built] building MIR for `polynomial::plot::<impl at src/polynomial/plot.rs:9:1: 60:2>::print`
#1 [unsafety_check_result] unsafety-checking `polynomial::plot::<impl at src/polynomial/plot.rs:9:1: 60:2>::print`
end of query stack
error: could not compile `polynomial`
```

Project 3:
Clippy issues one linter warning (out of multiple), but fails to compile the project.
```
Checking complex_numbers v0.1.0 (/data/data/com.termux/files/home/projects/rust/complex_numbers)
warning: unused import: `super::compare_f64::cmp_f64`
  --> src/complex/polar.rs:16:5
   |                                                              16 | use super::compare_f64::cmp_f64;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^                                 |
   = note: `#[warn(unused_imports)]` on by default

error: internal compiler error: compiler/rustc_mir_build/src/build/scope.rs:1397:17: cannot unwind from goto -> bb9
  --> src/complex/cartesian.rs:90:9
   |
90 | /         if *self == Self::new(T::zero(), T::zero()) {
91 | |             panic!("cannot divide by zero")
92 | |         }
   | |_________^

thread 'rustc' panicked at 'Box<dyn Any>', /home/builder/.termux-build/rust/src/compiler/rustc_errors/src/lib.rs:1093:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.1.56

query stack during panic:
#0 [mir_built] building MIR for `complex::cartesian::<impl at src/complex/cartesian.rs:16:1: 115:2>::reciprocal`
#1 [unsafety_check_result] unsafety-checking `complex::cartesian::<impl at src/complex/cartesian.rs:16:1: 115:2>::reciprocal`
end of query stack
warning: `complex_numbers` (bin "complex_numbers") generated 1 warning
error: could not compile `complex_numbers`; 1 warning emitted
```

Project 4:
```
Checking geometric_algebra v0.1.0 (/data/data/com.termux/files/home/projects/rust/geometric_algebra)
error: internal compiler error: compiler/rustc_mir_build/src/build/scope.rs:1397:17: cannot unwind from goto -> bb73
   --> src/ga_3d/bivector.rs:195:13                                   |
195 | /             if self.zx != 0.0 {
196 | |                 if self.zx.abs() != 1.0 {                 197 | |                     string.push_str(&self.zx.to_string());
198 | |                 }
...   |
202 | |                 string.push_str("zx");
203 | |             }
    | |_____________^

thread 'rustc' panicked at 'Box<dyn Any>', /home/builder/.termux-build/rust/src/compiler/rustc_errors/src/lib.rs:1093:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.1.56

query stack during panic:
#0 [mir_built] building MIR for `ga_3d::bivector::<impl at src/ga_3d/bivector.rs:165:1: 207:2>::fmt`
#1 [unsafety_check_result] unsafety-checking `ga_3d::bivector::<impl at src/ga_3d/bivector.rs:165:1: 207:2>::fmt`
end of query stack
error: could not compile `geometric_algebra`
```

Project 5:
Like in project 3, clippy issues one warning, but still fails to compile.
```
Checking hello_cargo v0.1.0 (/data/data/com.termux/files/home/projects/rust/fraction)
warning: unused import: `std::f64::consts`
 --> src/main.rs:6:5
  |
6 | use std::f64::consts;
  |     ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error: internal compiler error: compiler/rustc_mir_build/src/build/mod.rs:842:17: no terminator on block 3
  --> src/fraction.rs:23:5
   |
23 | /     pub fn from_i64(num: i64, den: i64) -> Self {
24 | |         let sign_num = num >= 0;
25 | |         let sign_den = den >= 0;
26 | |         let num = if sign_num { num } else { -num };
...  |
29 | |         Self::new(num as u64, den as u64, sign)
30 | |     }
   | |_____^

thread 'rustc' panicked at 'Box<dyn Any>', /home/builder/.termux-build/rust/src/compiler/rustc_errors/src/lib.rs:1093:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.1.56

query stack during panic:
#0 [mir_built] building MIR for `fraction::<impl at src/fraction.rs:15:1: 97:2>::from_i64`
#1 [unsafety_check_result] unsafety-checking `fraction::<impl at src/fraction.rs:15:1: 97:2>::from_i64`
end of query stack
warning: `hello_cargo` (bin "hello_cargo") generated 1 warning
error: could not compile `hello_cargo`; 1 warning emitted
```

### What steps will reproduce the bug?

Running cargo clippy inside a rust project.

### What is the expected behavior?

Clippy should lint and compile the project.

### System information

termux-info:

```
Application version:
0.113
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://packages.termux.org/apt/termux-main/ stable main
# science-repo (sources.list.d/science.list)
deb https://packages.termux.org/apt/termux-science science stable
# game-repo (sources.list.d/game.list)
deb https://packages.termux.org/apt/termux-games games stable
Updatable packages:
termux-tools/stable 0.146 all [upgradable from: 0.145]
Android version:
10
Kernel build information:
Linux localhost 4.9.179-perf+ #1 SMP PREEMPT Tue Jul 13 22:40:25 CST 2021 aarch64 Android
Device manufacturer:
OnePlus
Device model:
ONEPLUS A6003
```

### PR DESCRIPTION
**Pull request by xtkoba**
Supersedes #8711.
Fixes #7870

### CODE REVIEW

**Comment by thunder-coding** [comment]
Fun fact: You could have made use of Repology's advanced package search features to find out what versions of llvm are packaged in Ubuntu 20.04. https://repology.org/projects/?search=llvm&maintainer=&category=&inrepo=ubuntu_20_04&notinrepo=&repos=&families=&repos_newest=&families_newest=

**Comment by thunder-coding** [comment]
This also hopefully seems to fix #7870 as mentioned in https://github.com/rust-lang/rust-clippy/issues/7952#issuecomment-964760250

**Review by thunder-coding — APPROVED** [review]

File: scripts/setup-ubuntu.sh [hunk 1]
```diff
@@ -108,6 +108,11 @@ PACKAGES+=" composer"
 
 # Needed by package rust.
 PACKAGES+=" libssl-dev" # Needed to build Rust
+PACKAGES+=" llvm-12-dev"
+PACKAGES+=" llvm-12-tools"
+PACKAGES+=" clang-12"
+
+# Formerly needed by package rust. Not sure if needed by other packages.
 PACKAGES+=" clang-10"
```
Comment (buttaface) [inline scripts/setup-ubuntu.sh:116]: Yes, [Swift still needs this](https://github.com/termux/termux-packages/blob/master/packages/swift/build.sh#L107), I will try to remove it when I update to 5.5.3 this weekend.

Comment (buttaface) [inline scripts/setup-ubuntu.sh:116]: Yes, [Swift still needs this](https://github.com/termux/termux-packages/blob/master/packages/swift/build.sh#L107), I will try to remove it when I update to 5.5.3 this weekend.

Comment (buttaface) [inline scripts/setup-ubuntu.sh:116]: Yes, [Swift still needs this](https://github.com/termux/termux-packages/blob/master/packages/swift/build.sh#L107), I will try to remove it when I update to 5.5.3 this weekend.

Comment (xtkoba) [reply scripts/setup-ubuntu.sh:116]: I see. Thanks!

**Pull request merged**